### PR TITLE
Mines and Paint Gun

### PR DESCRIPTION
**Pull request by s-schmitt**
Here is both sets of mines and the Paint Gun that "vomits" projectiles. Keep in mind that the Paint Gun and Super Laser will have cooldowns soon.

https://mocsarcade.github.io/starjunk/92

### CODE REVIEW

File: source/Junkership.js [hunk 1]
```diff
@@ -7,6 +7,10 @@ import Reference from "./Reference.js"
 import Projectile from "./Projectile.js"
 import Score from "./Score.js"
 import Sound from "./Sound.js"
+import {PowerUp, PeaShoota, TriShoota, FiveShoota, RapidFire, RapidSprayShot,
+    SprayShot, SuperSprayShot, CrazySprayShot, VertSprayShot,
+    VertShoota, BFG, Laser, PiercingLaser, SuperLaser, Mine,
+    SuperMine, PaintShot} from "./PowerUp.js"
```
Comment (codaventure) [inline source/Junkership.js:13]: You don't actually need this or `this.WeaponList` anymore as I removed those dependencies a few PR's ago. :)

File: source/Projectile.js [hunk 2]
```diff
@@ -51,12 +67,31 @@ export default class Projectile extends Pixi.Sprite {
...
   onCollision(collidedWith) {
-      if(this.piercing != true)
+      if(this.projectileType == "mine") {
+          this.onHit.explodeEnemy(this)
+          this.destroy()
+      }
+      if(this.projectileType == "superMine") {
+          this.onHit.explodeEnemy(this)
+          this.destroy()
+      }
```
Comment (codaventure) [inline source/Projectile.js]: Actually, I was right in what I said in Slack. You're only calling `this.destroy()`. You still need to call `collidedWith.destroy()` to remove the trashbot, as `this.destroy()` only removes the projectile.

**Comment by s-schmitt** [comment]
https://mocsarcade.github.io/starjunk/92.f

**Comment by Phyoog** [comment]
Thoughts...

Mines need to be bigger. It's hard to connect with them. Maybe make the smalls the size of the large and the larges a bit bigger than that. They need to have some sort of lifespan too, the screen can get super-cluttered with them _way_ too easily and the spinning makes it look hyper busy. Also, the larger mines at least need to actually have an explosion that radiates out on impact and kills anything caught in the explosion. For the smaller mines, I can see it as the explosion is only big enough to kill one ship, but the larger really needs to have an AoE.

Lastly, all of the mines need to be destroyed when the gameOver function is called.

Still haven't seen the paint gun yet, so I can't give an opinion on that. I'll keep testing.

**Comment by Phyoog** [comment]
Why don't you make the PaintGun the starting weapon in the next build you push so it can be easily tested?

**Comment by s-schmitt** [comment]
Will do

**Comment by codaventure** [comment]
I don't know why, but I really like the paint gun. I guess it looks different from most of the other powerups, but is still surprisingly effective.

**Comment by codaventure** [comment]
@Binome Regarding things needing destroyed during gameOver function call, shouldn't junk also be destroyed? The current iteration preserves the junk after gameOver.

**Comment by s-schmitt** [comment]
I am putting a 5 second despawn time on the mines for now

**Comment by ehgoodenough** [comment]
How about a five mine max, placing anymore removes the first one.

File: source/Projectile.js [hunk 3]
```diff
@@ -68,16 +68,22 @@ export default class Projectile extends Pixi.Sprite {
           }
       }
 
+      if(this.despawn == true) {
```
Comment (Binome) [inline source/Projectile.js]: This strikes me as a bad way to do this, it means you're checking an attribute that is only defined for mines and superMines. Trying to compare an undeclared variable raised my hackles, and is usually undefined behavior. If this code only executes if you're dealing with a mine or superMine, why don't you just check `projectileType` instead? Shouldn't this work just fine with `if(this.projectileType === "mine" || this.projectileType === "superMine")`

**Comment by Phyoog** [comment]
@codaventure Good point. I'll take care of that.

An upper limit on mines sounds like a better idea than a despawn time.

File: source/Projectile.js [hunk 4]
```diff
@@ -68,7 +68,7 @@ export default class Projectile extends Pixi.Sprite {
           }
       }
 
-      if(this.despawn == true) {
+      if(this.mineDespawn == true) {
```
Comment (Phyoog) [inline source/Projectile.js]: This strikes me as a bad way to do this, it means you're checking an attribute that is only defined for mines and superMines. Trying to compare an undeclared variable raised my hackles, and is usually undefined behavior. If this code only executes if you're dealing with a mine or superMine, why don't you just check projectileType instead? Shouldn't this work just fine with if(this.projectileType === "mine" || this.projectileType === "superMine")

**Comment by s-schmitt** [comment]
Here is a new build that limits the amount of mines: https://mocsarcade.github.io/starjunk/92.g

**Comment by Phyoog** [comment]
New build? (You don't need to give it a new name, you can just overwrite 92)

**Comment by HolySchmitt** [comment]
https://mocsarcade.github.io/starjunk/92

**Comment by s-schmitt** [comment]
https://mocsarcade.github.io/starjunk/92

**Comment by Phyoog** [comment]
Paint gun and lasers seem fine.

File: source/Projectile.js [hunk 5]
```diff
@@ -103,10 +103,16 @@ export default class Projectile extends Pixi.Sprite {
...
+          this.onHit.explodePlayer(this)
           var curMine = this.shotBy.mineArray.indexOf(this)
           delete this.shotBy.mineArray[curMine]
           this.destroy()
+          this.explosionRadius = new Pixi.Rectangle(
+              this.x + 1 ,
+              this.y + 1 ,
+              5 ,
+              5 )
+          this.explosionRadius.destroy()
```
Comment (Phyoog) [inline source/Projectile.js]: You're creating the explosion radius and then immediately destroying it. So nothing is actually getting killed by being in the blast zone. You'll need to check the rectangle against `Trashbot.Inventory` for collisions and destroy any trashbots in the radius.

**Comment by Phyoog** [comment]
I see explosion radius got put in and then taken out. What's the plan for that?

**Comment by s-schmitt** [comment]
Still working on it

**Comment by Phyoog** [comment]
Okay. Well, what we've got is a concrete step and I've yet to uncover any serious bugs. I'm going to merge what we've got and let you get back to it.

**Pull request merged**